### ISSUE TITLE
loss_spatial: difference between code implementation and discription in paper

### ISSUE THREAD

**Issue by SMY19999** [google-research/readout_guidance#11]
Thanks for your great contribution to bringing a new conditional method to Diffusion-based generation.

For loss_spatial, in Section 4.1 of the paper, you say "_We find that our method works well with extracted features from the **unconditional branch**, using the empty string “” as the prompt for all images._ "

I find the corresponding codes in readout_guidance/rg_operators.py:

```python
def loss_spatial(obs_feat, gt_feat, edit, latents_scale=(0, 1)):
    obs_feat = einops.rearrange(obs_feat, 'b w h c -> b c w h')    # cond branch
    gt_feat = einops.rearrange(gt_feat, 'b w h c -> b c w h')         # uncond branch
    control_feat = edit["control"][None, ...].detach().clone()
    control_feat = control_feat.to(obs_feat.dtype)
    _, _, obs_w, obs_h = obs_feat.shape
    # Normalize from arbitrary output scale
    obs_feat = renormalize(obs_feat, edit["control_range"], latents_scale)
    gt_feat = renormalize(gt_feat, edit["control_range"], latents_scale)
    control_feat = renormalize(control_feat, edit["control_range"], latents_scale)
    control_feat = torch.nn.functional.interpolate(control_feat, (obs_w, obs_h))
    loss = torch.nn.functional.mse_loss(obs_feat, control_feat, reduction="none")  # Loss
    loss = loss.mean()
    return loss, obs_feat, gt_feat
```
May I ask if the content I annotated, **cond branch** and **uncond branch**, is correct？

If I'm right, the loss is computed with **obs_feat (cond branch)**, which is different from the paper.

Look forward to hearing back from you.

**Comment by g-luo** [comment]
Hi @SMY19999, thanks for your question. 

- We use the unconditional branch only during **training**. You can see this is the case when you look at the [training code](https://github.com/google-research/readout_guidance/tree/main/readout_training), where the prompt is set to the empty string "" in the [config](https://github.com/google-research/readout_guidance/blob/ba3d7611b9299e25bfba5410a3a4180443772c03/readout_training/configs/train_spatial.yaml#L25), which is equivalent to using the unconditional branch.
- We indeed use the conditional branch for **guidance**, as you have noticed in the codebase.

I apologize if Sec 4.1 was confusing; the more clear statement would be "We only train the readout head on unconditional features, but at sampling time it is also able to guide conditional features."

**Comment by SMY19999** [comment]
Thanks for your reply, this helps me and others to understand your contribution better!

**Comment by GloryyrolG** [comment]
> Hi @SMY19999, thanks for your question.
> 
> * We use the unconditional branch only during **training**. You can see this is the case when you look at the [training code](https://github.com/google-research/readout_guidance/tree/main/readout_training), where the prompt is set to the empty string "" in the [config](https://github.com/google-research/readout_guidance/blob/ba3d7611b9299e25bfba5410a3a4180443772c03/readout_training/configs/train_spatial.yaml#L25), which is equivalent to using the unconditional branch.
> * We indeed use the conditional branch for **guidance**, as you have noticed in the codebase.
> 
> I apologize if Sec 4.1 was confusing; the more clear statement would be "We only train the readout head on unconditional features, but at sampling time it is also able to guide conditional features."

hi @g-luo et al.,

`feats = torch.cat([uncond, cond])`, so aren't we calc the loss for both uncond[1] (=feats[1]) & cond[1] (=feats[b // 2 + 1]) instead of only cond? thx & best,

https://github.com/google-research/readout_guidance/blob/main/readout_guidance/rg_pipeline.py#L167

**Comment by g-luo** [comment]
Hi @GloryyrolG, yes you are correct; the above statement should be revised to "We indeed use the conditional (and unconditional branch) for **guidance**" as linked in the above code.
